numpy/ma/core.py

```diff
@@ -6184,32 +6184,43 @@ def isMaskedArray(x):
 
 
 class MaskedConstant(MaskedArray):
-    # We define the masked singleton as a float for higher precedence.
-    # Note that it can be tricky sometimes w/ type comparison
-    _data = data = np.array(0.)
-    _mask = mask = np.array(True)
-    _baseclass = ndarray
+
+    __singleton = None
 
     def __new__(self):
-        return self._data.view(self)
+        if self.__singleton is None:
+            # We define the masked singleton as a float for higher precedence.
+            # Note that it can be tricky sometimes w/ type comparison
+            data = np.array(0.)
+            mask = np.array(True)
+            self.__singleton = MaskedArray(data, mask=mask).view(self)
 
-    def __array_finalize__(self, obj):
-        return
+        return self.__singleton
 
-    def __array_prepare__(self, obj, context=None):
-        return self.view(MaskedArray).__array_prepare__(obj, context)
+    def __array_finalize__(self, obj):
+        if self.__singleton is None:
+            # this handles the `.view` in __new__, which we want to copy across
+            # properties normally
+            return super(MaskedConstant, self).__array_finalize__(obj)
+        elif self is self.__singleton:
+            # not clear how this can happen, play it safe
+            pass
+        else:
+            # everywhere else, we want to downcast to MaskedArray, to prevent a
+            # duplicate maskedconstant.
+            self.__class__ = MaskedArray
+            self.__array_finalize__(obj)
```

Comment:
It would seem `anything.view(MaskedConstant)` is almost certainly an error....  And setting `__class__` is quite awful. But I defer to your judgment - I'm on holidays and worry looking in more detail is not going to be a short process...

(and I agree we cannot change `__array_finalize__`)